examples/Example3/particle.cc

```diff
@@ -0,0 +1,250 @@
+	cudaSafe(cudaMemcpyToSymbol(d_ntotal, &ntotal, sizeof(int)));
+
+	/* create space for newly created particles */
+	newtot = nparticles;
+	cudaSafe(cudaMalloc((void **)&ptr, nparticles * sizeof(particle)));
+	cudaSafe(cudaMemcpyToSymbol(d_new_particles, &ptr, sizeof(ptr)));
+	cudaSafe(cudaMemcpyToSymbol(d_newcur, &nalive, sizeof(int)));
+	cudaSafe(cudaMemcpyToSymbol(d_newtot, &newtot, sizeof(int)));
+#endif
+}
+
+void particle_free()
+{
+#ifndef __CUDACC__
+	particles.clear();
+#else
+	ntotal = 0;
+	nalive = 0;
+	particle *ptr = nullptr;
+
+	/* free space for particles */
+	cudaSafe(cudaMemcpyFromSymbol(&ptr, d_particles, sizeof(ptr)));
+	cudaSafe(cudaFree(ptr));
+
+	/* free space for new particles */
+	cudaSafe(cudaMemcpyFromSymbol(&ptr, d_new_particles, sizeof(ptr)));
+	cudaSafe(cudaFree(ptr));
+#endif
+}
+
+void particle_resize(int nparticles)
+{
+#ifndef __CUDACC__
+	particles.reserve(nparticles);
+#else
+	/* no shrinking */
+	if (nparticles <= ntotal)
+		return;
+
+	ntotal = nparticles;
+	particle *old_ptr = nullptr, *new_ptr = nullptr;
+	cudaSafe(cudaMalloc((void **)&new_ptr, nparticles * sizeof(particle)));
+	cudaSafe(cudaMemcpyFromSymbol(&old_ptr, d_particles, sizeof(old_ptr)));
+	cudaSafe(cudaMemcpy((void **)&new_ptr, old_ptr, ntotal * sizeof(particle),
+			    cudaMemcpyDeviceToDevice));
+	cudaSafe(cudaMemcpyToSymbol(d_particles, &new_ptr, sizeof(new_ptr)));
+	cudaSafe(cudaMemcpyToSymbol(d_ntotal, &ntotal, sizeof(int)));
+	cudaSafe(cudaFree(old_ptr));
+	cudaSafe(cudaDeviceSynchronize());
+#endif
+}
+
+#if 0
+__device__ void new_particle_resize(int nparticles)
+{
+	particle *old_ptr = d_new_particles;
+	particle *new_ptr = malloc(nparticles * sizeof(particle));
+	memcpy(new_ptr, old_ptr, d_newtot * sizeof(particle));
+	d_newtot = nparticles;
+	d_new_particles = new_ptr;
+	free(old_ptr);
+}
+#endif
+
+#ifdef __CUDACC__
+static __global__ void reset_new()
+{
+	if (threadIdx.x == 0 && blockIdx.x == 0 && d_newcur < 0)
+		d_newcur = 0;
+}
+
+void reset_new_particle_count()
+{
+	reset_new<<<1, 1>>>();
+	cudaCheckKernelCall();
+}
+
+__global__ void fetch_new_particles_from_buffer()
+{
+	for (int id = blockIdx.x * blockDim.x + threadIdx.x; id < d_ntotal;
+	     id += blockDim.x * gridDim.x) {
+		/* replace dead particle with new particle */
+		if (!d_particles[id].alive) {
+			int newid = atomicSub(&d_newcur, 1) - 1;
+
+			/* no more new particles to fetch */
+			if (newid < 0)
+				break;
+
+			debug_printf("moving particle (%d)\n", newid);
+			d_particles[id] = d_new_particles[newid];
+			track_init(d_particles[id]);
+		}
+	}
+}
+#endif
+
+#ifdef __CUDACC__
+static __global__ void update_number_of_particles_alive()
+{
+	__shared__ int nalive_s;
+
+	if (threadIdx.x == 0 && blockIdx.x == 0)
+		d_nalive = 0;
+
+	if (threadIdx.x == 0)
+		nalive_s = 0;
+
+	for (int id = blockIdx.x * blockDim.x + threadIdx.x; id < d_ntotal;
+	     id += blockDim.x * gridDim.x) {
+		if (d_particles[id].alive)
+			atomicAdd(&nalive_s, 1);
+	}
+
+	__syncthreads();
```

Comment:
Note that above, threads are writing atomically to the shared variable, `native_s`, and below, thread 0 is reading from that variable to accumulate into `d_nalive`. Therefore, thread 0 needs to wait for all others to have modified `nalive_s` before it can read the final value and add it to `d_nalive`. If I didn't add `__syncthreads()` at this point, thread 0 can read the state of `nalive_s` before all threads have written to it and get the total number of particles alive wrong. However, you are right that it may be necessary to add another `__syncthreads()` before the for loop as well, although given the fetching of global memory that has to happen before incrementing `nalive_s`, I thought it was safe not to add an additional `__syncthreads()` there, and so far I have not observed any problems due to that being missing (but have observed problems if the existing call to `__syncthreads()` is removed).